export type { Primitive, Props, PropValue, State, StateView } from "./shared"

btakita opened the pull request:
- mini-van.d.ts
- van-plate.d.ts

Removed duplicate types.

Tao-VanJS reviewed (commented):
For now the duplication is intended. I want to make Mini-Van easy to use in a sense that people can just download one `.js` file and one `.d.ts` file to start with, in addition to use Mini-Van with NPM. IMHO, the amount of duplication is tolerable.

If you have good reasons to believe otherwise, I would be happy to know :-)

btakita commented:
I'm adding some isomorphic helpers. Consistent types are useful for adding code that is meant to work both on the browser & server. The second commit adds more types into `shared`.

btakita commented:
`shared.ts` can instead import from `mini-van.d.ts` & `van-plate.d.ts`, allowing the two files to have no dependencies.

Tao-VanJS commented:
Hi,

Thanks for your commit!

Overall this PR seems to do many things and make the type system of Mini-Van much more complicated. But I'm struggling to understand the main motivation of the PR. Preferably, if there are multiple objectives to achieve, we can break down into smaller PRs, so that the change for each objective becomes more clear.

Here are my general opinion regarding the PR:

1. For now, a good feature of Mini-Van is: you can download a single `.js` file and a single `.d.ts` file to start working with, without the need of NPM integration. I would like to keep that feature, unless there is convincing benefit if we give up on it. I tried to find some TypeScript bundling solutions but couldn't get a workable one. In the end, I decided to just have some duplicate code among `mini-van.d.ts`, `van-plate.d.ts` and `shared.ts`. For now the amount of duplication is manageable. I think we can keep the duplication until the complexity hits an inflection point.
2. I noticed that one improvement of this PR is to enable more type-checkings in `shared.ts` for tag functions. I can see the benefit. If this is something you want to improve, I would prefer to have a minimal PR just to address that. This way it will be easier to weigh in the pros and cons of the PR. Ideally, I'm hoping we can make the improvement without breaking the feature in 1).

btakita commented:
I reverted `mini-van.d.ts` & `van-plate.d.ts`, added common types into `shared.ts`, & rebased.

Here is a couple of extensions that I'm working on:

```ts
const protoOf = Object.getPrototypeOf
const objProto = protoOf({})
export type PropsWithUndefined = Record<string, PropValue|undefined|StateView<PropValue>|(()=>PropValue)>
export type TagFuncWithUndefined = (first?:PropsWithUndefined|ChildDom, ...rest:readonly ChildDom[])=>Element
export function Raw(van:VanObj, html:string):ChildDom {
	if (globalThis['window']) {
		const frag = document.createDocumentFragment()
		const temp = document.createElement('div')
		temp.innerHTML = html
		while (temp.firstChild) {
			frag.appendChild(temp.firstChild)
		}
		return frag as any as ChildDom
	}
	else {
		const { __proto__ } = van.tags.div()
		return {
			__proto__,
			renderToBuf(buf:any[]) {
				buf.push(html)
			},
		} as any as ChildDom
	}
}
export function vanProxy<V extends VanObj>(van:V):Omit<V, 'tags'>&{
	tags:Readonly<Record<string, TagFuncWithUndefined>>
} {
	let tags = new Proxy<typeof van.tags>(van.tags, {
		get(target, p, receiver) {
			const f = Reflect.get(target, p, receiver)
			if (!f) return f
			const fn:typeof van.tags[string] = (...args)=>{
				let [props, ...children] = protoOf(args[0] ?? 0) === objProto ? args : [{}, ...args]
				return f(cleanProps(props), ...children)
			}
			return fn
		}
	})
	return new Proxy<typeof van>(van, {
		get(target, p, receiver) {
			if (p === 'tags') return tags
			return Reflect.get(target, p, receiver)
		}
	})
}
function cleanProps(props:Props) {
	for (const key of Object.keys(props)) {
		if (props[key] === undefined) delete props[key]
	}
	return props
}
```

Tao-VanJS commented on src/shared.ts:
```diff
@@ -1,26 +1,46 @@
+type NodeType<ElementType extends HasFirstChild> =
+  Omit<ElementType["firstChild"], "after" | "before" | "remove" | "replaceWith">
+
+export type ValidChildDomValue<ElementType = any, TextNodeType = any> =
```
Why having the type parameters default to `any`?

btakita replied:
I'm using the examples I provided to guide development. `van-plate.d.ts` `ValidChildDomValue` is not a generic type while `mini-van.d.ts` `ValidChildDomValue` has two generics. `any` is not a requirement. A placeholder default type could be used instead. There may be a better pattern for the type, but this is what I came up with to solve the type issue in the [examples](https://github.com/vanjs-org/mini-van/pull/8#issuecomment-1788093557).

btakita replied:
I think `any` is too permissive. I will put more work into this.

Tao-VanJS replied:
The typical use case of `shared.ts` is for shared components between SSR and CSR, with the following pattern for example:

```js
const Component(van: VanObj) {
  ...
}
```

For this PR to provide better type checking than status-quo, the default `ElementType` and `TextNodeType` need to have meaningful type information to enable the checking.

I was trying to solve this problem, but found it is indeed very hard to provide a meaningful `ElementType` and `TextNodeType` that are shared between VanJS, `van-plate` and `mini-van`. Thus I gave up on it and chose to enforce just some basic type checking in `shared.ts`.

Tao-VanJS commented on test/mini-van.test.ts:
```diff
@@ -131,7 +131,7 @@ const runTests = (van: Van, msgDom: Element) => {
         p({
           "data-index": state1,
-          "data-id": () => van.val(state2) + 2,
+          "data-id": () => van.val<number>(state2) + 2,
```
Do you need to change this line?

btakita replied:
No. I'll rebase it out.

btakita commented:
@Tao-VanJS I took a different approach with `shared.ts`, adding a `VanShape` type. `VanShaped` is used in the generics for types where `mini-van.d.ts` & `van-plate.d.ts` differ. Some types were added to `mini-van.d.ts` & `van-plate.d.ts` as well.

[This gist](https://gist.github.com/btakita/dfd966bd612fec82824d937357e46412) contains the example that I used.

btakita commented:
I left in `VanObj` but I think `VanShape` could replace `VanObj`.

Tao-VanJS commented:
Hi @btakita,

Thank you so much for your contribution! I did take a in-depth review of this PR. It's a difficult decision but I haven't come to a point for merging this PR in.

First, about my view on TypeScript and `.d.ts` files in VanJS-series projects: I am considering TypeScript as add-on type annotations. The core library should work on its own without the `.d.ts` annotations, and the purpose of `d.ts` annotations is to try our best to assist our users for writing correct client code while keeping the type annotations simple and easy to understand. While advanced TypeScript features allow us to do very sophisticated type gymnastics but doing too much will make the type system hard to digest and navigate.

For this specific PR, I'm afraid that we're pushing the complexity of relevant TypeScript files to a point that is no longer easy to understand. This can be judged by the size increases of the files, especially in `shared.ts`. It's really hard for me to be sure that the benefit we can get from this PR worths the complexity.

Moreover, I don't think this specific PR can significantly enhance the type-checking capabilities of Mini-Van, especially with the introduction of `any` in different places. Here are a few issues that I noticed:

1. In Line 20 of `van-plate.d.ts`:
```ts
export type BindingFunc = (dom: Element) => ValidChildDomValue
```
I don't think this change is right. Indeed, the signature of `BindingFunc` is quite complex, which is why I'm using `any` for `van-plate` to avoid making the type too complex in `van-plate`. In VanJS, the signature is:
```ts
export type BindingFunc = ((dom?: Node) => ValidChildDomValue) | ((dom?: Element) => Element)
```
In `mini-van` the signature is:
```ts
export type BindingFunc<ElementType extends HasFirstChild, TextNodeType> =
  | ((dom?: ElementType | TextNodeType) => ValidChildDomValue<ElementType, TextNodeType>)
  | ((dom?: ElementType) => ElementType)
```
2. In Line 49 of `van-plate.d.ts`:
```ts
readonly add: AddFunc<any, any>
```
This is reducing the type checking capability of `van-plate`, as literally any function is compatible to `AddFunc<any, any>`.
3. In `shared.ts`, the definition of `VanShape`, I noticed that the `mini-van` portion of the definition:
```ts
{
  readonly _BindingFunc?: MiniBindingFunc<any, any>
  readonly _ChildDom?: MiniChildDom<any, any>
  readonly _ValidChildDomValue?: MiniValidChildDomValue<any, any>
  readonly add: MiniAddFunc<any, any>
  readonly html: MiniHtmlFunc<any, any>
  readonly tags: MiniTags<any, any>
  readonly tagsNS: MiniTagsNSFunc<any, any>
}
```
Because of the using of `any`, we're literally making any function compatible with these defined types, thus the `VanShape` has hardly any type-checking capability.

It's not 100% clear to me what you're trying to achieve in this PR. i.e.: what exactly is missing in the current type annotations of Mini-Van for the things that you plan to work on. If you're trying to create some utility functions that can work for both `mini-van` and `van-plate`, my suggestion is to start with something simple with just some basic type-checking capabilities, and leverage weak-typing (such as `any`) when type gymnastics are pushed for too far. Then we can start to gradually introduce more type-checking, step by step, with each step we make sure to keep the complexity under control. Another approach is for utility functions to deal with more specific concrete types. The types in `van-plate` are pretty simple and specific. The types in `mini-van` are quite generic but you can start working with the concrete specializations that you care about.

Let me know if it makes sense to you.

btakita commented:
Hi @Tao-VanJS, thanks for taking a look. I'm attempting to create isomorphic functions to extend Van. I ran into type issues & usability issues using the type system on `main`. The example I posted in the gist has to work for both the `van-plate` & `mini-van`.

I'm having trouble with the types in the `main` branch, particularly around the types that have the `<ElementType extends HasFirstChild, TextNodeType>` generics in `mini-van`. I hope to have a shared set of types that can provide a better experience to verify both type systems work.

# Domain Consistency

First I had to wrap my head around the terminology. `van-plate` for SSR & `mini-van` for CSR. When making isomorphic components, I want to use the `shared` types as much as possible, as it indeed is shared between SSR & CSR. Using SSR types for isomorphic components limits the type system capabilities when the component is rendered in CSR. I don't think I could even figure out how to do the example in the gist with the type system in `main`. Isomorphic components would have to reconcile & support both the `<ElementType extends HasFirstChild, TextNodeType>` generic for `mini-van` & the lack of generics for `van-plate`.

Making `van-plate` types do double duty as a generic type for isomorphic components adds complexity & influences the decision of how to design the types in `van-plate`. The types in `van-plate` are no longer meant to accurately model what happens in SSR, but to also be a generic type for isomorphic components. This forces types to be overly-wide for SSR.

You pointed out:

> ```ts
> export type BindingFunc = (dom: Element) => ValidChildDomValue
> ```
> I don't think this change is right. Indeed, the signature of BindingFunc is quite complex, which is why I'm using any for van-plate to avoid making the type too complex in van-plate. In VanJS, the signature is:
> ```ts
> export type BindingFunc = ((dom?: Node) => ValidChildDomValue) | ((dom?: Element) => Element)
> ```

I updated the type to:

```ts
export type BindingFunc = (dom?: Element) => ValidChildDomValue
```

From what I understand, `dom?: Element` is the actual type expected in `van-plate` for `BindingFunc`. If there are other types expected, why not explicitly model what is happening during execution? If `van-plate`'s `BindingFunc` double as the common isomorphic type, then we have to use `dom:any`.

> In Line 49 of van-plate.d.ts:

I missed removing the generic. `AddFunc` is not a generic type in `van-plate`.

# Consistency with Generics

`VanShape` is the current try to keep `mini-van` & `van-plate` simple with no dependencies, while providing a consistent type api to target isomorphic code.

> Because of the using of any, we're literally making any function compatible with these defined types, thus the VanShape has hardly any type-checking capability.

Good catch! I strengthened `VanShape`:

```ts
type VanCommon = {
  readonly state: <T>(initVal: T) => State<T>
  readonly val: <T>(s: T|StateView<T>) => T
  readonly oldVal: <T>(s: T|StateView<T>) => T
  readonly derive: <T>(f: () => T) => State<T>
  readonly _: UnderscoreFunc
}
export type VanShape = VanCommon&(MiniVan|VanPlate)
```
and removed the defaults for generics using `VanShape` (`<V extends VanShape>` instead of `<V extends VanShape = VanPlate>`):

Now `VanShape` has stronger typing, there were a couple of errors in the example code that I had to fix.

I updated the [example gist](https://gist.github.com/btakita/dfd966bd612fec82824d937357e46412).

## Raw component

I had to use a ReturnType ternary to resolve correctly.

```ts
type Return<V extends VanShape> = V extends MiniVan ? DocumentFragment : VanPlateElement
```

There are tests at the bottom of the file.

## vanCleanProps

With stronger type checking, I got a type error when trying:

```ts
				let [
					props,
					...children
				] = (protoOf(args[0] ?? 0) === objProto ? args : [{}, ...args]) as [Props, ...ChildDom<V>[]]
				return tag(cleanProps(props), ...children)
```

It seems like `TagFunc<V>` does not narrow the type when checking for valid valid arguments. It needs to work for both `MiniVan` & the `VanPlate` types even if the ternary type check is the same between `TagFunc<V>` & `ChildDom<V>`. To fix it, I had to use `ChildDom<VanPlate>`. I'm not satisfied with this solution, but at least it's a stronger type check than what was originally in the `shared` `VanObj`, which was:

```ts
readonly tags: Record<string, Function>
```

Note the type return by `vanCleanProps`

```ts
export type PropsWithUndefined = Record<string, Props[string]|undefined>
export type TagFuncWithUndefined<V extends VanShape> =
	(first?:Props|ChildDom<V>|PropsWithUndefined, ...rest:readonly ChildDom<V>[])=>ReturnType<TagFunc<V>>
export type VanWithUndefined<V extends VanShape> = V&{
	tags:Record<string, TagFuncWithUndefined<V>>
	tagsNS:(namespaceURI:string)=>Readonly<Record<string, TagFuncWithUndefined<V>>>
}
```

I'm not sure how do accomplish the same thing without chaos with the `main` branch...

btakita commented:
I realize a couple of things now. `van` is meant to run in the browser. `mini-van` is meant to run on the server to render to a DOM library (e.g. JSDOM) or in the browser with limited prop & reactive support. `van-plate` is meant to render to a html string. Is this correct?

I also realize that `BindingFunc` is not run in `van-plate`, so `(dom: any)` makes sense.

With that being said, isomorphic types that delegate based on the Van type would need to depend on `van-core` in addition to this library. That makes me think that it would make more sense to not include the delegation in this library. The 3 type sets can support the delegation use case by keeping the same set of names.

Tao-VanJS commented:
> I realize a couple of things now. van is meant to run in the browser. mini-van is meant to run on the server to render to a DOM library (e.g. JSDOM) or in the browser with limited prop & reactive support. van-plate is meant to render to a html string. Is this correct?

Yes. Your understanding is correct.

I will try to explain my thought process regarding type systems in VanJS-series projects.

Generally speaking, for type systems of projects with the level of generic and flexibility like VanJS, designing a good type system is a challenging task and needs juggling among the 3 main type of problems:

1. Types are too loose. Cost of this type of problems:
  * We're not able to catch some bugs before run time.
  * We lose some perks from the IDE, especially, a better suggestion for auto-complete.
2. Types are too tight. Cost of this type of problems:
  * Correct client code result in type errors.
3. Types are too complex. Cost of this type of problems:
  * Type system is hard to understand. It's hard to figure out whether there are bugs in the type system by reading the code.
  * Type system is hard to deal with. It's hard to maintain and keep evolving the type system.
  * When users get type errors, the error messages are usually hard to understand, sometimes misleading. This is quite evident in [C++ template errors](https://www.reddit.com/r/ProgrammerHumor/comments/dtquhf/c_template_errors/).

There are times where it's hard to avoid all 3 types of problems altogether. Thus in case we need to do some tradeoffs, in my opinion, the stack ranking of importance is: 3 > 2 > 1.

This is why in `shared.ts`, we chose to have problem 1, rather than problem 3 or 2.

Another remark is that, when we're developing VanJS, `mini-van`, `van-plate` and VanX, because of the generic and flexibility we have in the implementations, it will make the type system more complex than desired if we had them implemented in TypeScript. As a result, we chose to have the JavaScript-based implementation first, and then try to annotate the public interface of these projects in `d.ts` files with a type system that is reasonably simple and tries to catch client-side bugs as much as possible. Separating the implementation and type annotations helps us decouple the entangled problems, and as a result greatly simplify things. Indeed, we recognize the limitation of the type checking thus that's why we have `van.debug.js` which enables more sanity checks than what are possible in TypeScript.

For your specific work stream, I would recommend a similar approach. If I understand correctly, what you want to have is an extension of `van` object with your preferred behavior. You want the extension to be generic enough so that it can cover VanJS, `mini-van` and `van-plate`. I think by trying to implement your extension completely in JavaScript first and then add a `d.ts` file to annotate the public interface can simplify lots of things.

I think you can simply have some public interface like:

```ts
export function extend<T extends ClientVan | MiniVan<Element, Text> | VanPlate>(van: T): T
```

For `Raw` function, you can have something like:

```ts
type RawReturn<T extends ClientVan | MiniVan<Element, Text> | VanPlate> =
  T extends ClientVan ? DocumentFragment :
  T extends MiniVan<Element, Text> ? MiniVanDocumentFragment :
  PlateElement

export function Raw<T extends ClientVan | MiniVan<Element, Text> | VanPlate>(van: T, html: string): RawReturn<T>
```

Does it make sense to you?

btakita commented:
That makes sense. I agree that `.js` + `.d.ts` simplifies things for library development while `.ts` has been good for app development. I like how having `.js` decoupled from the types allows "hidden" behavior while the types more-or-less represent the API of the library.

I appreciate you bringing up the 3 tradeoffs re: types. I think different development contexts weigh the tradeoffs differently from each other.

I have a large number of codebases to maintain, both open source & proprietary. Not many developers (if any) use these libraries. I experiment with some techniques & have had significant churn with the code with new libraries, techniques. I'm more concerned about managing isomorphic development across many codebases. I will need to be able to come back to a codebase after time passes *(sometimes over a year later)* & understand the code. My code bases have also gone through considerable development churn as the state of the art changes, better libraries are introduced, etc.

That being said, I favor flat architectures, consistent type systems, injecting context, reactive programming, modular global state management, custom code for simple behavior, simple dependencies backed by a skillful community for more complex behavior.

I'm not as concerned about some complexity in the types, assuming there is a coherent set of patterns to the types. I'm more concerned about balancing the 3 points depending on the development context. I don't think I can codify a rigid order of preference of the 3 tradeoffs you mentioned at this point.

That being said, I moved the types into https://github.com/btakita/van-type-delegate. I converted my website over to use VanJS. I left in the SolidJS components as well for comparison.

I have to say that I'm very satisfied with moving the components over to VanJS. I think it integrates very well with Astro. For example, within a *.astro or *.mdx component, I can use `<Fragment>` to render the VanJS component. It's nice to use plain old functions. A VanJS component in Astro does not support `<slot>`, but it's not a big loss. Using Astro `<slot>` in a framework (e.g. SolidJS) component has complicated progressive async rendering behavior.

I have a `V_footnote` & `V_footnote_list` components where the footnote html is stored & rendered in the `V_footnote_list` at the bottom of the page. The VanJS implementation is cleaner & more coherent than the Astro + SolidJS (or any other supported framework) implementation...Due to the simplicity of calling synchronous functions instead of working with progressive async rendering.

I'm tweaking van-type-delegate to make it as smooth as possible while providing adequate isomorphic type checking.

Tao-VanJS commented:
Thanks @btakita for sharing your context and perspective! And also thanks for your kind remark on VanJS. I'm really glad it provides you the developer experience you enjoy.

Regarding the support of asynchronous rendering, we have [`Await` utility component](https://vanjs.org/vanui#await) in VanJS, thanks to @Hunter-Gu's contribution.
